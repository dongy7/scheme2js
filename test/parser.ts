import Compiler from '../src/compiler'

it('parses empty program', () => {
  const tokens = Compiler.tokenizer('')
  const parser = new Compiler.parser(tokens)
  expect(parser.parse()).toMatchSnapshot()
})

it('parses function call', () => {
  const tokens = Compiler.tokenizer('(add 1 2)')
  const parser = new Compiler.parser(tokens)
  expect(parser.parse()).toMatchSnapshot()
})

it('parses nested function call', () => {
  const tokens = Compiler.tokenizer('(add 1 (sub 2 3))')
  const parser = new Compiler.parser(tokens)
  expect(parser.parse()).toMatchSnapshot()
})

it('parses expr with operator', () => {
  const tokens = Compiler.tokenizer('(+ 1 (- 2 3))')
  const parser = new Compiler.parser(tokens)
  expect(parser.parse()).toMatchSnapshot()
})

it('parses define expression', () => {
  const tokens = Compiler.tokenizer('(define a 1)')
  const parser = new Compiler.parser(tokens)
  expect(parser.parse()).toMatchSnapshot()
})

it('parses if expression', () => {
  const tokens = Compiler.tokenizer('(if (eq 1 2) 3 4)')
  const parser = new Compiler.parser(tokens)
  expect(parser.parse()).toMatchSnapshot()
})

it('parses nested if expression', () => {
  const tokens = Compiler.tokenizer('(if (eq 1 2) (add 4 5) 6)')
  const parser = new Compiler.parser(tokens)
  expect(parser.parse()).toMatchSnapshot()
})
