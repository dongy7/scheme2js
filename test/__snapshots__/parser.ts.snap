// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses define expression 1`] = `
Object {
  "body": Array [
    Object {
      "ref": Object {
        "type": "SymbolLiteral",
        "value": "a",
      },
      "type": "DefineExpr",
      "value": Object {
        "type": "NumberLiteral",
        "value": "1",
      },
    },
  ],
  "type": "Program",
}
`;

exports[`parses empty program 1`] = `
Object {
  "body": Array [],
  "type": "Program",
}
`;

exports[`parses expr with operator 1`] = `
Object {
  "body": Array [
    Object {
      "params": Array [
        Object {
          "type": "NumberLiteral",
          "value": "1",
        },
        Object {
          "params": Array [
            Object {
              "type": "NumberLiteral",
              "value": "2",
            },
            Object {
              "type": "NumberLiteral",
              "value": "3",
            },
          ],
          "proc": Object {
            "type": "SymbolLiteral",
            "value": "-",
          },
          "type": "CallExpression",
        },
      ],
      "proc": Object {
        "type": "SymbolLiteral",
        "value": "+",
      },
      "type": "CallExpression",
    },
  ],
  "type": "Program",
}
`;

exports[`parses function call 1`] = `
Object {
  "body": Array [
    Object {
      "params": Array [
        Object {
          "type": "NumberLiteral",
          "value": "1",
        },
        Object {
          "type": "NumberLiteral",
          "value": "2",
        },
      ],
      "proc": Object {
        "type": "SymbolLiteral",
        "value": "add",
      },
      "type": "CallExpression",
    },
  ],
  "type": "Program",
}
`;

exports[`parses if expression 1`] = `
Object {
  "body": Array [
    Object {
      "alt": Object {
        "type": "NumberLiteral",
        "value": "4",
      },
      "conseq": Object {
        "type": "NumberLiteral",
        "value": "3",
      },
      "test": Object {
        "params": Array [
          Object {
            "type": "NumberLiteral",
            "value": "1",
          },
          Object {
            "type": "NumberLiteral",
            "value": "2",
          },
        ],
        "proc": Object {
          "type": "SymbolLiteral",
          "value": "eq",
        },
        "type": "CallExpression",
      },
      "type": "IfExpr",
    },
  ],
  "type": "Program",
}
`;

exports[`parses nested function call 1`] = `
Object {
  "body": Array [
    Object {
      "params": Array [
        Object {
          "type": "NumberLiteral",
          "value": "1",
        },
        Object {
          "params": Array [
            Object {
              "type": "NumberLiteral",
              "value": "2",
            },
            Object {
              "type": "NumberLiteral",
              "value": "3",
            },
          ],
          "proc": Object {
            "type": "SymbolLiteral",
            "value": "sub",
          },
          "type": "CallExpression",
        },
      ],
      "proc": Object {
        "type": "SymbolLiteral",
        "value": "add",
      },
      "type": "CallExpression",
    },
  ],
  "type": "Program",
}
`;

exports[`parses nested if expression 1`] = `
Object {
  "body": Array [
    Object {
      "alt": Object {
        "type": "NumberLiteral",
        "value": "6",
      },
      "conseq": Object {
        "params": Array [
          Object {
            "type": "NumberLiteral",
            "value": "4",
          },
          Object {
            "type": "NumberLiteral",
            "value": "5",
          },
        ],
        "proc": Object {
          "type": "SymbolLiteral",
          "value": "add",
        },
        "type": "CallExpression",
      },
      "test": Object {
        "params": Array [
          Object {
            "type": "NumberLiteral",
            "value": "1",
          },
          Object {
            "type": "NumberLiteral",
            "value": "2",
          },
        ],
        "proc": Object {
          "type": "SymbolLiteral",
          "value": "eq",
        },
        "type": "CallExpression",
      },
      "type": "IfExpr",
    },
  ],
  "type": "Program",
}
`;
